src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp

```diff
@@ -1041,18 +1045,48 @@ void BlockLocalPositionEstimator::correctGps() {	// TODO : use another other met
 void BlockLocalPositionEstimator::correctVision() {
 
 	math::Vector<6> y;
+
+	static float last_vision_x = 0.0f;
+	static float last_vision_y = 0.0f;
+	static float last_vision_z = 0.0f;
+
 	y(0) = _sub_vision_pos.get().x - _visionHome(0);
 	y(1) = _sub_vision_pos.get().y - _visionHome(1);
 	y(2) = _sub_vision_pos.get().z - _visionHome(2);
-	if(_visionVelInitialized) {
+
+	if(_visionVelInitialized) {	// If subscribed to valid vision_speed_estimate topic, use its data
 		y(3) = _sub_vision_vel.get().x - _visionBaseVel(0);
 		y(4) = _sub_vision_vel.get().y - _visionBaseVel(1);
 		y(5) = _sub_vision_vel.get().z - _visionBaseVel(2);
 	}
-	else {
-		y(3) = 0.0f;
-		y(4) = 0.0f;
-		y(5) = 0.0f;
+	else {	// Else, derivate velocity from position
```

Comment:
The Kalman filter basically does this for you already. If you don't have a vision velocity estimate, you should just modify the C and R matrices to only have the 3 position measurements.